server/app/services/grid_service_deployer.rb

```diff
@@ -39,53 +39,67 @@ def selected_nodes
     nodes
   end
 
+  # @return [Boolean]
   def deploy
-    info "starting to deploy #{self.grid_service.to_path}"
-    self.grid_service_deploy.set(:deploy_state => :ongoing)
     deploy_rev = Time.now.utc.to_s
+    info "starting to deploy #{self.grid_service.to_path} at #{deploy_rev}"
+    self.grid_service_deploy.set(:deploy_state => :ongoing)
     self.grid_service.set(:deployed_at => deploy_rev)
 
-    deploy_futures = []
     total_instances = self.instance_count
     self.grid_service.grid_service_instances.where(:instance_number.gt => total_instances).destroy
+
+    deploy_futures = []
     total_instances.times do |i|
       instance_number = i + 1
       unless self.grid_service.reload.deploying?
-        raise "halting deploy of #{self.grid_service.to_path}, desired state has changed"
+        raise DeployError, "desired state has changed"
       end
-      self.deploy_service_instance(total_instances, deploy_futures, instance_number, deploy_rev)
+
+      deploy_futures << self.deploy_service_instance(instance_number, deploy_rev)
+
+      pending_deploys = deploy_futures.select{|f| !f.ready?}
+      if pending_deploys.size >= (total_instances * self.min_health).floor || pending_deploys.size >= 20
+        info "throttling service instance #{self.grid_service.to_path} deploy because of min_health limit (#{pending_deploys.size} instances in-progress)"
+        sleep 0.1 until pending_deploys.any?{|f| f.ready? }
+      end
+
+      # bail out early if anything fails
+      if deploy_futures.select{|f| f.ready? }.map{|f| f.value }.any?{|d| d.error? } # raises on any Future exceptions
+        raise DeployError, "one or more instances failed"
```

Comment:
Imho this could be a deploy_opt, cancel or continue on error. Calculation based on min_health is too magical.